src/object/srv_obj_migrate.c

```diff
@@ -0,0 +1,1722 @@
+{
+	tls->mpt_refcount++;
+}
+
+void
+migrate_pool_tls_put(struct migrate_pool_tls *tls)
+{
+	tls->mpt_refcount--;
+	if (tls->mpt_refcount == 0)
+		migrate_pool_tls_destroy(tls);
+}
+
+struct migrate_pool_tls_create_arg {
+	uuid_t	pool_uuid;
+	uuid_t	pool_hdl_uuid;
+	uuid_t  co_hdl_uuid;
+	d_rank_list_t *svc_list;
+	uint64_t max_eph;
+	int	version;
+};
+
+int migrate_pool_tls_create_one(void *data)
+{
+	struct migrate_pool_tls_create_arg *arg = data;
+	struct obj_tls			   *tls = obj_tls_get();
+	struct migrate_pool_tls		   *pool_tls;
+	int rc;
+
+	pool_tls = migrate_pool_tls_lookup(arg->pool_uuid, arg->version);
+	if (pool_tls != NULL) {
+		/* Some one else already created, because collective function
+		 * might yield xstream.
+		 */
+		migrate_pool_tls_put(pool_tls);
+		return 0;
+	}
+
+	D_ALLOC_PTR(pool_tls);
+	if (pool_tls == NULL)
+		return -DER_NOMEM;
+
+	uuid_copy(pool_tls->mpt_pool_uuid, arg->pool_uuid);
+	uuid_copy(pool_tls->mpt_poh_uuid, arg->pool_hdl_uuid);
+	uuid_copy(pool_tls->mpt_coh_uuid, arg->co_hdl_uuid);
+	pool_tls->mpt_version = arg->version;
+	pool_tls->mpt_pool_hdl = DAOS_HDL_INVAL;
+	pool_tls->mpt_rec_count = 0;
+	pool_tls->mpt_obj_count = 0;
+	pool_tls->mpt_size = 0;
+	pool_tls->mpt_generated_ult = 0;
+	pool_tls->mpt_executed_ult = 0;
+	pool_tls->mpt_root_hdl = DAOS_HDL_INVAL;
+	pool_tls->mpt_max_eph = arg->max_eph;
+	pool_tls->mpt_pool = ds_pool_child_lookup(arg->pool_uuid);
+	d_list_add(&pool_tls->mpt_list, &tls->ot_pool_list);
+
+	pool_tls->mpt_refcount = 1;
+	rc = daos_rank_list_copy(&pool_tls->mpt_svc_list, arg->svc_list);
+	D_DEBUG(DB_TRACE, "TLS %p create for "DF_UUID" ver %d rc %d\n",
+		pool_tls, DP_UUID(pool_tls->mpt_pool_uuid), arg->version, rc);
+	return rc;
+}
+
+static struct migrate_pool_tls*
+migrate_pool_tls_lookup_create(struct ds_pool *pool, int version,
+			       uuid_t pool_hdl_uuid, uuid_t co_hdl_uuid,
+			       uint64_t max_eph)
+{
+	struct migrate_pool_tls *tls = NULL;
+	struct migrate_pool_tls_create_arg arg = { 0 };
+	daos_prop_t		*prop;
+	struct daos_prop_entry	*entry;
+	int			rc = 0;
+
+	tls = migrate_pool_tls_lookup(pool->sp_uuid, version);
+	if (tls)
+		return tls;
+
+	D_ALLOC_PTR(prop);
+	if (prop == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	rc = ds_pool_iv_prop_fetch(pool, prop);
+	if (rc)
+		D_GOTO(out, rc);
+
+	entry = daos_prop_entry_get(prop, DAOS_PROP_PO_SVC_LIST);
+	D_ASSERT(entry != NULL);
+
+	uuid_copy(arg.pool_uuid, pool->sp_uuid);
+	uuid_copy(arg.pool_hdl_uuid, pool_hdl_uuid);
+	uuid_copy(arg.co_hdl_uuid, co_hdl_uuid);
+	arg.version = version;
+	arg.max_eph = max_eph;
+	arg.svc_list = (d_rank_list_t *)entry->dpe_val_ptr;
+	rc = dss_task_collective(migrate_pool_tls_create_one, &arg, 0,
+				 DSS_ULT_REBUILD);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to create migrate tls: %d\n",
+			DP_UUID(pool->sp_uuid), rc);
+		D_GOTO(out, rc);
+	}
+
+	/* dss_task_collecitve does not do collective on xstream 0 */
+	rc = migrate_pool_tls_create_one(&arg);
```

Comment:
just that there are several xxx_one, maybe can think more to refine the definition/name later.